lib/firenest/replicated_state/store.ex

```diff
@@ -0,0 +1,124 @@
+defmodule Firenest.ReplicatedState.Store do
+  @moduledoc false
+
+  defstruct [:values, :pids]
+
+  # Common data exchange format:
+  # {{key, pid}, value}
+
+  def new(name) do
+    values = :ets.new(name, [:named_table, :protected, :ordered_set, read_concurrency: true])
+    pids = :ets.new(__MODULE__.Pids, [:private, :duplicate_bag, keypos: 2])
+
+    %__MODULE__{values: ets_whereis(values), pids: pids}
+  end
+
+  def list(%__MODULE__{values: values}, key) do
+    local = {{{key, :_}, :"$1", :_}, [], [:"$1"]}
+    remote = {{{key, :_}, :"$1"}, [], [:"$1"]}
+    :ets.select(values, [local, remote])
+  end
+
+  def list_local(%__MODULE__{values: values}) do
+    ms = [{{:"$1", :"$2", :_}, [], [{{:"$1", :"$2"}}]}]
+    :ets.select(values, ms)
+  end
+
+  def present?(%__MODULE__{values: values}, key, pid) do
+    ets_key = {key, pid}
+    :ets.member(values, ets_key)
+  end
+
+  def fetch(%__MODULE__{values: values}, key, pid) do
+    ets_key = {key, pid}
+
+    case :ets.match(values, {ets_key, :"$1", :"$2"}) do
+      [[value, local_delta]] -> {:ok, value, local_delta}
+      [] -> :error
+    end
+  end
+
+  def local_put(%__MODULE__{values: values, pids: pids} = state, key, pid, value, local_delta) do
+    ets_key = {key, pid}
+
+    true = :ets.insert_new(values, {ets_key, value, local_delta})
+    :ets.insert(pids, ets_key)
+    state
+  end
+
+  def local_delete(%__MODULE__{values: values, pids: pids} = state, key, pid) do
+    ets_key = {key, pid}
+    ms = [{ets_key, [], [true]}]
+
+    case :ets.select_delete(pids, ms) do
```

Comment:
Is this equivalent to ets:delete_object/2?

Reply:
Oh, delete_object does not say if it was deleted or not. :(

Reply:
Yes, but it allows us to know if we actually deleted or not. `:ets.delete_object/2` does not give this information back.